utils/zerovec/src/ule/mod.rs

```diff
@@ -51,69 +61,80 @@ where
 
     /// Validates a byte slice, `&[u8]`.
     ///
-    /// If `Self` is not well-defined for all possible bit values, the bytes should be validated,
-    /// that they can be transumted to a `Self` and `Self::Error` should be returned otherwise.
-    fn validate_byte_slice(_bytes: &[u8]) -> Result<(), Self::Error>;
+    /// If `Self` is not well-defined for all possible bit values, the bytes should be validated.
+    /// If the bytes can be transmuted to a valid slice of `Self`, then `Ok` should be returned;
+    /// otherwise, `Self::Error` should be returned.
+    fn validate_byte_slice(bytes: &[u8]) -> Result<(), Self::Error>;
 
     /// Parses a byte slice, `&[u8]`, and return it as `&[Self]` with the same lifetime.
     ///
     /// If `Self` is not well-defined for all possible bit values, the bytes should be validated,
     /// and `Self::Error` should be returned if they are not valid.
     ///
+    /// The default implementation executes [`Self::validate_byte_slice`] followed by a length
+    /// check and then [`Self::from_byte_slice_unchecked`].
+    ///
     /// # Safety
     ///
     /// Implementations of this method may involve `unsafe{}` blocks to cast the pointer to the
     /// correct type. It is up to the implementation to reason about the safety. Keep in mind that
     /// `&[Self]` and `&[u8]` may have different lengths.
     ///
-    /// Implementors must return a slice to the same region of memory if parsing succeeds.
-    ///
-    /// Ideally, implementations call [`ULE::from_byte_slice_unchecked()`] after validation.
+    /// If successful, this function must return the same result as
+    /// [`Self::from_byte_slice_unchecked`], which is a slice to the same region of memory.
     ///
-    /// The default implementation executes `Self::validate_byte_slice` followed by
-    /// `Self::from_byte_slice_unchecked`.
-    fn parse_byte_slice(bytes: &[u8]) -> Result<&[Self], Self::Error> {
-        Self::validate_byte_slice(bytes)?;
+    /// This method can be overridden with a fully-safe implementation if possible.
+    fn parse_byte_slice(bytes: &[u8]) -> Result<&[Self], UleError<Self::Error>> {
+        Self::validate_byte_slice(bytes).map_err(|e| UleError::InvalidBytes(e))?;
+        if bytes.len() % mem::size_of::<Self>() != 0 {
+            return Err(UleError::InvalidLength);
+        }
```

Comment:
`question (n-b)`: In my patches I put this test in `validate_byte_slice`. This addition means that you don't want the check of length being divisible by size be in `validate_byte_slice`?

Reply:
It's easy to check the invariant in the default impl, and relying on validate_byte_slice is prone to error.

Reply:
I made a separate error type with the check in `parse_byte_slice` because:

1. It's not possible to verify that `validate_byte_slice` performs this check (debug_assert only works if there is a code path)
2. We *can* perform the check automatically, so it seems like we *should*
3. It's easy to miss the check in code review

Reply:
Right, but "it's not possible to verify X" is the standard state of _all_ unsafe code. My experience with Rust unsafe is that it's far better to consolidate the unsafe requirements in one well-documented place. This is basically splitting out the responsibilities and making it trickier to reason about: now `validate_byte_slice` is in charge of validating the slice only in _some_ ways.

"We can perform this check so we should" is why I'm suggesting debug asserts, we should absolutely be sprinkling the code with more of these.

I think it's table stakes for Rust unsafe code review for folks to be verifying that the function is defined according to the checklist, if reviewers aren't doing that that's a much bigger problem.

Perhaps we should more strongly require comments that ensure the checklists are satisfied. We require safety explainer comments, but in the case of `unsafe trait`s, we could require checklists and require that these checklists be proven for each impl.

Reply:
Okay, I'm convinced.  I changed it to a bullet point in the checklist.

Reply:
It turns out that there is another problem here.  What error type should PlainOldULE and and CharULE return when the lengths are not correct?  In my previous draft, I had them returning UleError (as the default), but now that I've gotten rid of that, I don't know what error they should return.  In an effort to not block this otherwise productive PR with documentation changes, I filed a follow-up issue for this: https://github.com/unicode-org/icu4x/issues/1144